src/model.coffee

```diff
@@ -1017,14 +1021,18 @@ exports.SocketStartToken = class SocketStartToken extends StartToken
 exports.SocketEndToken = class SocketEndToken extends EndToken
   constructor: (@container) -> super; @type = 'socketEnd'
 
-  stringify: ->
-    if @prev is @container.start or
+  stringify: (opts = DEFAULT_STRINGIFY_OPTS) ->
+    # If preserveEmpty is turned on, substitute our placeholder string
+    # for an empty socket
+    if opts.preserveEmpty and @prev is @container.start or
         @prev.type is 'text' and @prev.value is ''
       return @container.emptyString
+
+    # Otherwise, do nothing, and allow the socket to stringify to ''
     else ''
 
 exports.Socket = class Socket extends Container
-  constructor: (@emptyString, @precedence = 0, @handwritten = false, @classes = [], @dropdown = null) ->
+  constructor: (@emptyString, @precedence = 0, @handwritten = false, @classes = [], @dropdown = null, @parseContext = null) ->
```

Comment:
Here you're adding a new field to Socket and it strikes me that they all need to be documented.
Can you break this out to one field per line and add a comment explaining the purpose of each one, including the new one parseContext?